cpp/oneapi/dal/algo/decision_forest/backend/gpu/infer_model_manager.hpp

```diff
@@ -0,0 +1,198 @@
+        daal::algorithms::decision_forest::classification::internal::ModelImpl;
+    using daal_model_ptr_t = daal::algorithms::decision_forest::classification::ModelPtr;
+};
+
+template <>
+struct daal_types_map<task::regression> {
+    using daal_tree_impl_t = daal::algorithms::dtrees::internal::TreeImpRegression<>;
+    using daal_model_impl_t = daal::algorithms::decision_forest::regression::internal::ModelImpl;
+    using daal_model_ptr_t = daal::algorithms::decision_forest::regression::ModelPtr;
+};
+
+using namespace daal::algorithms::dtrees::internal;
+using namespace daal::services::internal;
+
+template <typename Float, typename Index, typename Task = task::by_default>
+class infer_model_manager {
+    using infer_context_t = infer_context<Float, Index, Task>;
+    using model_t = model<Task>;
+    using model_impl_t = detail::model_impl<Task>;
+    using model_interop_impl_t =
+        model_interop_impl<typename daal_types_map<Task>::daal_model_ptr_t>;
+    using daal_model_impl_t = typename daal_types_map<Task>::daal_model_impl_t;
+    using daal_model_ptr_t = typename daal_types_map<Task>::daal_model_ptr_t;
+    using daal_decision_tree_table_t = daal::algorithms::dtrees::internal::DecisionTreeTable;
+    using daal_decision_tree_node_t = daal::algorithms::dtrees::internal::DecisionTreeNode;
+
+public:
+    explicit infer_model_manager(const cl::sycl::queue& q,
+                                 const infer_context_t& ctx,
+                                 const model_t& model)
+            : queue_(q),
+              ctx_(ctx) {
+        const daal_model_impl_t* const daal_model_ptr = get_daal_model(model);
+
+        ONEDAL_ASSERT(dal::detail::integral_cast<size_t>(ctx_.tree_count_) ==
+                      daal_model_ptr->size());
+
+        const auto tree_count = ctx_.tree_count_;
+
+        std::vector<const daal_decision_tree_table_t*> tree_list;
+        tree_list.resize(tree_count);
+
+        size_t tree_size_max = 0;
+        for (Index i = 0; i < tree_count; ++i) {
+            tree_list[i] = daal_model_ptr->at(i);
+            tree_size_max = std::max(tree_size_max, tree_list[i]->getNumberOfRows());
+        }
+
+        if (tree_size_max > dal::detail::limits<Index>::max()) {
+            throw domain_error(dal::detail::error_messages::input_model_tree_has_invalid_size());
+        }
+
+        max_tree_size_ = dal::detail::integral_cast<Index>(tree_size_max);
+
+        dal::detail::check_mul_overflow(max_tree_size_, tree_count);
+        const Index tree_block_size = max_tree_size_ * tree_count;
+
+        auto fi_list_host = dal::backend::primitives::ndarray<Index, 1>::empty({ tree_block_size });
+        auto lc_list_host = dal::backend::primitives::ndarray<Index, 1>::empty({ tree_block_size });
+        auto fv_list_host = dal::backend::primitives::ndarray<Float, 1>::empty({ tree_block_size });
+
+        Index mul_class_count_and_tree_in_group_count =
+            ctx_.class_count_ * ctx.tree_in_group_count_;
```

Comment:
`check_mul_overflow` can be used this way.
```suggestion
            dal::detail::check_mul_overflow(ctx_.class_count_, ctx.tree_in_group_count_);
```

Reply:
applied